app/controllers/submissions_controller.rb

```diff
@@ -1,6 +1,19 @@
 class SubmissionsController < ApplicationController
-  skip_before_action :authenticate_user!, only: [:new, :create, :thank_you]
-  layout 'dashboard', only: [:valid, :rated, :to_rate, :rejected, :results]
+  skip_before_action :authenticate_user!, only: [:confirm, :new, :create, :thank_you]
+  layout 'dashboard', only: [:valid, :all, :rated, :to_rate, :rejected, :results]
+
+  def confirm
+    token = params.require(:confirmation_token)
+    submission = Submission.find_by!(confirmation_token: token)
+    if submission.past_confirmation_due_date?
+      render text: "Time for confirmation expired!"
+    else
+      submission.confirmed!
+      render text: "You confirmed your invitation!"
+    end
```

Comment:
Not necessarily, but personally I'd probably move those action to service. But as there are not so much of them, I don't have strong opinion about this.